RK3566 Bluetooth

ananjaser1211 opened the pull request:
* WIP to figure out why HID Devices can't register as input devices on RK3566


* Problem : 
  Turn on bluetooth and pair a controller (DS4 / 8bitdo)
  Device connects but does not register as an input device
  Unpair and try to connect an audio device
  Device connects with audio playback and control


* Currently testing on DS4, it still does not register as a HID, but so far changes have fixed auto disconnection on connect and a bunch of other errors

* Current connection log on Dualshock4 (connects / pairs)

```
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/plugin.c:init_plugin() System does not support bap plugin
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/plugin.c:init_plugin() System does not support bass plugin
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/plugin.c:init_plugin() System does not support mcp plugin
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/plugin.c:init_plugin() System does not support vcp plugin
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: profiles/audio/micp.c:micp_init() D-Bus experimental not enabled
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/plugin.c:init_plugin() System does not support micp plugin
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/plugin.c:init_plugin() System does not support ccp plugin
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/plugin.c:init_plugin() System does not support csip plugin
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: Bluetooth management interface 1.22 initialized
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: src/adapter.c:reset_adv_monitors_complete() Failed to reset Adv Monitors: Failed (0x03)
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: Battery Provider Manager created
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: Failed to clear UUIDs: Failed (0x03)
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: Failed to add UUID: Failed (0x03)
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: Failed to add UUID: Failed (0x03)
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: Failed to add UUID: Failed (0x03)
......
Apr 12 15:11:05 RK3566-X55 bluetoothd[13384]: PHY setting rejected for 0: Invalid Parameters
Apr 12 15:11:11 RK3566-X55 bluetoothd[13384]: profiles/input/device.c:control_connect_cb() connect to 1C:A0:B8:64:AB:5D: Host is down (112)
Apr 12 15:11:17 RK3566-X55 bluetoothd[13384]: profiles/input/device.c:control_connect_cb() connect to 1C:A0:B8:64:AB:5D: Host is down (112)
Apr 12 15:11:23 RK3566-X55 bluetoothd[13384]: profiles/input/device.c:control_connect_cb() connect to 1C:A0:B8:64:AB:5D: Host is down (112)
Apr 12 15:11:24 RK3566-X55 bluetoothd[13384]: Authorization request for non-connected device!?
Apr 12 15:15:26 RK3566-X55 bluetoothd[13384]: No matching connection for device
Apr 12 15:15:26 RK3566-X55 bluetoothd[13384]: src/device.c:search_cb() 1C:A0:B8:64:AB:5D: error updating services: Connection reset by peer (104)
```

bomberstudios commented:
Would it be possible to have access to a build to test these changes? I’m still struggling with setting up a build environment but I’m more than happy to help with testing

magicmyth commented:
Just a bit of testing feedback. I've tried the branch with a Powkiddy RGB30 and 8BitDo SN30 pro and still cannot connect (well technically it does then immediately fails) using branch or dev. Below is my output from bluetoothctl:

<details>
<summary>bluetoothctl output:</summary>

```
# bluetoothctl 
Agent registeredct to bluetoothd...
hci0 E4:17:D8:0F:F7:7B type BR/EDR connect failed (status 0x04, Connect Failed)
[bluetooth]# list
Controller 98:03:CF:64:8F:22 S-POW [default]
hci0 E4:17:D8:0F:F7:7B type BR/EDR connect failed (status 0x04, Connect Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
[8Bitdo SN30 Pro]# list
Controller 98:03:CF:64:8F:22 S-POW [default]
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
[bluetooth]# devices
Device E4:17:D8:0F:F7:7B 8Bitdo SN30 Pro
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
[bluetooth]# discoverable on
hci0 new_settings: powered connectable bondable ssp br/edr le secure-conn 
hci0 new_settings: powered connectable discoverable bondable ssp br/edr le secure-conn 
Changing discoverable on succeeded
[CHG] Controller 98:03:CF:64:8F:22 Discoverable: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connect failed (status 0x04, Connect Failed)
[CHG] Device E4:17:D8:0F:F7:7B Trusted: no
[DEL] Device E4:17:D8:0F:F7:7B 8Bitdo SN30 Pro
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[NEW] Device E4:17:D8:0F:F7:7B 8Bitdo SN30 Pro
hci0 type 7 discovering on
[CHG] Controller 98:03:CF:64:8F:22 Discovering: yes
[NEW] Device 59:04:2A:DA:07:70 Kitchen
hci0 new_link_key E4:17:D8:0F:F7:7B type 0x04 pin_len 0 store_hint 1
hci0 device_flags_changed: E4:17:D8:0F:F7:7B (BR/EDR)
     supp: 0x00000001  curr: 0x00000000
[CHG] Device E4:17:D8:0F:F7:7B Bonded: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 3
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connect failed (status 0x04, Connect Failed)
[CHG] Device E4:17:D8:0F:F7:7B Paired: yes
[CHG] Device E4:17:D8:0F:F7:7B Trusted: yes
hci0 type 7 discovering off
[CHG] Device 59:04:2A:DA:07:70 RSSI is nil
[CHG] Controller 98:03:CF:64:8F:22 Discovering: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connect failed (status 0x04, Connect Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR connect failed (status 0x04, Connect Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[CHG] Device E4:17:D8:0F:F7:7B Modalias: usb:v045Ep02E0d0903
hci0 device_flags_changed: E4:17:D8:0F:F7:7B (BR/EDR)
     supp: 0x00000001  curr: 0x00000001
[CHG] Device E4:17:D8:0F:F7:7B UUIDs: 00001124-0000-1000-8000-00805f9b34fb
[CHG] Device E4:17:D8:0F:F7:7B UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: yes
[CHG] Device E4:17:D8:0F:F7:7B WakeAllowed: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 type 7 discovering on
[CHG] Controller 98:03:CF:64:8F:22 Discovering: yes
hci0 type 7 discovering off
[CHG] Controller 98:03:CF:64:8F:22 Discovering: no
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
[DEL] Device 59:04:2A:DA:07:70 Kitchen
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B ServicesResolved: no
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
[bluetooth]# discoverable on
Changing discoverable on succeeded
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 type 7 discovering on
[CHG] Controller 98:03:CF:64:8F:22 Discovering: yes
[CHG] Device E4:17:D8:0F:F7:7B RSSI: 0xffffffce (-50)
hci0 type 7 discovering off
[CHG] Device E4:17:D8:0F:F7:7B RSSI is nil
[CHG] Controller 98:03:CF:64:8F:22 Discovering: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 type 7 discovering on
[CHG] Controller 98:03:CF:64:8F:22 Discovering: yes
hci0 type 7 discovering off
[CHG] Controller 98:03:CF:64:8F:22 Discovering: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 new_settings: powered connectable bondable ssp br/edr le secure-conn 
[CHG] Controller 98:03:CF:64:8F:22 Discoverable: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 type 7 discovering on
[CHG] Controller 98:03:CF:64:8F:22 Discovering: yes
[CHG] Device E4:17:D8:0F:F7:7B RSSI: 0xffffffd2 (-46)
hci0 type 7 discovering off
[CHG] Device E4:17:D8:0F:F7:7B RSSI is nil
[CHG] Controller 98:03:CF:64:8F:22 Discovering: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 22
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 type 7 discovering on
[CHG] Controller 98:03:CF:64:8F:22 Discovering: yes
hci0 type 7 discovering off
[CHG] Controller 98:03:CF:64:8F:22 Discovering: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
[8Bitdo SN30 Pro]# discoverable on
hci0 new_settings: powered connectable bondable ssp br/edr le secure-conn 
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 new_settings: powered connectable discoverable bondable ssp br/edr le secure-conn 
Changing discoverable on succeeded
[CHG] Controller 98:03:CF:64:8F:22 Discoverable: yes
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
[bluetooth]# list
Controller 98:03:CF:64:8F:22 S-POW [default]
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
[bluetooth]# pair E4:17:D8:0F:F7:7B
Attempting to pair with E4:17:D8:0F:F7:7B
Failed to pair: org.bluez.Error.AlreadyExists
[bluetooth]# devices
Device E4:17:D8:0F:F7:7B 8Bitdo SN30 Pro
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
hci0 E4:17:D8:0F:F7:7B type BR/EDR disconnected with reason 2
[CHG] Device E4:17:D8:0F:F7:7B Connected: no
hci0 E4:17:D8:0F:F7:7B type BR/EDR connected eir_len 17
[CHG] Device E4:17:D8:0F:F7:7B Connected: yes
hci0 E4:17:D8:0F:F7:7B auth failed with status 0x05 (Authentication Failed)
```

</details>

Are there any other logs I can pull from that might be helpful for you?

Note I rebased this branch on top of dev before building.

ananjaser1211 commented:
> Just a bit of testing feedback. I've tried the branch with a Powkiddy RGB30 and 8BitDo SN30 pro and still cannot connect (well technically it does then immediately fails) using branch or dev. Below is my output from bluetoothctl:
> 
> bluetoothctl output:
> Are there any other logs I can pull from that might be helpful for you?
> 
> Note I rebased this branch on top of dev before building.

Ive been running tests on bluez 5.70 and 5.73 but pretty much both end up failing at HIDP

with the latest pushed branch im able to keep the device connected again, but HIDP is not working.

i also added `UserspaceHID=false` which fixed this issue 

```
Apr 19 19:48:42 RK3566-X55 bluetoothd[833]: profiles/input/device.c:input_device_register() bt_uhid_new_default: failed
Apr 19 19:48:42 RK3566-X55 bluetoothd[833]: src/service.c:service_probe() input-hid profile probe failed for 1C:A0:B8:64:AB:5D
```

this is after doing modprobe hidp , prior to that it would complain about socket not working.

```
Apr 19 20:58:55 RK3566-X55 bluetoothd[29022]: profiles/input/device.c:ioctl_is_connected() Can't get HIDP connection info
Apr 19 20:59:00 RK3566-X55 bluetoothd[29022]: profiles/input/device.c:control_connect_cb() connect to 1C:A0:B8:64:AB:5D: Host is down (112)
Apr 19 21:00:34 RK3566-X55 bluetoothd[29022]: profiles/input/device.c:ioctl_is_connected() Can't get HIDP connection info
Apr 19 21:00:35 RK3566-X55 bluetoothd[29022]: profiles/input/device.c:ioctl_is_connected() Can't get HIDP connection info
Apr 19 21:00:40 RK3566-X55 bluetoothd[29022]: profiles/input/device.c:control_connect_cb() connect to 1C:A0:B8:64:AB:5D: Host is down (112)
```

im still conflicted whether this is userspace or kernel, so im shuffling around alot of configs to see if anything helps.

r3claimer commented:
Some of these issues might have been resolved in 5.74 and 5.75. Might be worth trying to move forward and not back. 

https://github.com/LibreELEC/LibreELEC.tv/pull/8820

magicmyth commented:
@ananjaser1211 I tested out your latest commits and as you seemed to suggest it did not make any odds with getting my SN30 Pro to connect but it did make my Google Home devices show up which I had not noticed before (not that I attempted pairing with them. Just interesting).

Next I gave @r3claimer's suggestion of the latest Bluez 5.75 with some of the patches from LibreELEC (applied on dev) and the controller started working right away! It was already technically paired so I removed it and tried pairing again but that initially failed. So I restarted the system and for some reason bluetoothctl could no longer see the device controller but dmesg seemed to suggest it was still being picked up. I restarted again and this time the device showed up and when I attempted to pair to the SN30 Pro it paired right away with no issue. I had similar issues in the past with JELOS BSP kernel based setup so at least no regression there any more.

 If you guys would like I could make a branch and pull request with the Bluez 5.75 updates from LibreELEC but since @ananjaser1211 has put in the actual hard work here and seems to be doing a deeper dive in getting bluetooth working more reliable in general I'd rather see if they want to merge that into this branch.

ananjaser1211 commented:
> @ananjaser1211 I tested out your latest commits and as you seemed to suggest it did not make any odds with getting my SN30 Pro to connect but it did make my Google Home devices show up which I had not noticed before (not that I attempted pairing with them. Just interesting).
> 
> Next I gave @r3claimer's suggestion of the latest Bluez 5.75 with some of the patches from LibreELEC (applied on dev) and the controller started working right away! It was already technically paired so I removed it and tried pairing again but that initially failed. So I restarted the system and for some reason bluetoothctl could no longer see the device controller but dmesg seemed to suggest it was still being picked up. I restarted again and this time the device showed up and when I attempted to pair to the SN30 Pro it paired right away with no issue. I had similar issues in the past with JELOS BSP kernel based setup so at least no regression there any more.
> 
> If you guys would like I could make a branch and pull request with the Bluez 5.75 updates from LibreELEC but since @ananjaser1211 has put in the actual hard work here and seems to be doing a deeper dive in getting bluetooth working more reliable in general I'd rather see if they want to merge that into this branch.

Thats amazing to hear! did you apply any of the commits in this branch or just changed BT to 5.75 (with the patches) on dev ? i am still unable to pair my Dualshock4 controller, i will try reverting few things and see how it goes.

ananjaser1211 commented:
Update: the DS4 issue was caused by not having HID_PLAYSTATION and i needed to do `modprobe hidp` ! it works now, i will re-do the kernel config so we dont have to modprobe and i think the BT will be fine then!

magicmyth commented:
Excellent! Just to be clear, I updated Bluez 5.75 directly on dev with these patches from LibreElec:

* bluez-50-shared-uhid--Fix-crash-if-bt_uhid_destroy-free-replay.patch
* bluez-12-fix-obexd-after-5_73.patch
* bluez-22-Revert-build-Fix-distcheck.patch

And removed:

* bluez-06-AVRCP_TG_MDI_BV-04-C.patch
* bluez-12-fix-obexd-after-5_50.patch

I did _not_ use any patches/commits from _this_ branch (pull #102).

ananjaser1211 commented:
Branch has been forced pushed. Summary of changes :

- [bluez: update to 5.75](https://github.com/ROCKNIX/distribution/pull/102/commits/f2c0d01af1521e1fcdb6e7e2c92ac677b26cf39c)   * 5.75 is much more stable than either 5.72 and 5.73, 8bitDo works with it and allows for DS4 to work after few other patches
- [bluez: fix bt_uhid_new_default crash](https://github.com/ROCKNIX/distribution/pull/102/commits/e91408f57e796c255c0dfc4b7a1e2f2b6f3c2f7f)  | [blues: enable fastconnect](https://github.com/ROCKNIX/distribution/pull/102/commits/fc38375773ca7ccc7d7ff9aa8c657079a7e33c6b) * This fixes issues when connecting HID devices
- [bluez: trust sixaxis](https://github.com/ROCKNIX/distribution/pull/102/commits/77110385c649821b9b511ab5f979fae3b9450060) and subsequent updates fix an issue where DS4 Connects and disconnects due to pairing
- [RK3566: linux: HIDP needs to be a module](https://github.com/ROCKNIX/distribution/pull/102/commits/965eea7fcd2bda61c138c684ce71331544a1a179) | [bluez: autostart: modprobe hidp](https://github.com/ROCKNIX/distribution/pull/102/commits/94c349d2eee41ada67407d9abecc8fa1ea1cbc36) * HIDP was non functional and required to be a module, + modprobing it on startup finally allows DS4 to connect (and possibly other HID gamepads), without it the HIDP socket does not work and the device connects but is unusable
- [RK3566: bluetooth: add l2cap fix](https://github.com/ROCKNIX/distribution/pull/102/commits/7f8775d5c62bb0be9ea8466d9d2085c47ed0ffe1) | [RK3566: patch to fix "Align minimum encryption key size](https://github.com/ROCKNIX/distribution/pull/102/commits/7877e898695eb06f45a0981f8f69c5ceab9deabc) * Both fix issues i faced while debugging BT, the former should be removed when we rebase past 6.9-rc4 mainline according to github issue page
- [RK3566: enable HID_PLAYSTATION](https://github.com/ROCKNIX/distribution/pull/102/commits/e4308e13ef068480069bdfb02135040ee6426d44) * HID_SONY seems to be unstable for DS4, HID_PLAYSTATION and its dependencies fix that

TLDR; 
- With these changes im able to pair my Dualshock 4 controller, map it in ES and use it in Retroarch and flycast-sa (what i tested basically)
- While not all are mandatory, this is what was required to get my gamepad working as i dont have any other gamepads to verify with
- Pairing was done through the GUI and not CLI
- This should be tested on all other devices that use Bluetooth to insure i didn't break something that was working before

magicmyth commented:
Just realised I forgot to mention I tested ananjaser1211 latest updates and all is working just fine. Bluetooth is working with my 8bitDo and Gioteck WX4 at the same time (though the Gioteck has issues but they seem unrelated to the Bluetooth connection and existed for the BSP Jelos builds). Bluetooth seems real solid now. I'll try and find more bluetooth devices to test with but so far it seems at least on par if not better than the old BSP base.

sydarn commented:
Thanks for this works looks very promising, I didn't have time to test on my end yet. Hopefully tomorrow!

sydarn commented:
Tested with 8bitdo on rk3566 and rk3588.
